pkg/engine/utils/finegrained_exceptions.go

```diff
@@ -0,0 +1,228 @@
+package utils
+
+import (
+	"strings"
+
+	"github.com/go-logr/logr"
+	kyvernov2 "github.com/kyverno/kyverno/api/kyverno/v2"
+	"github.com/kyverno/kyverno/ext/wildcard"
+	engineapi "github.com/kyverno/kyverno/pkg/engine/api"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// MatchesFinegrainedException checks if the resource matches any fine-grained exception criteria
+// for the given policy rule. It returns true if the resource should be exempted based on
+// image-based or value-based exception criteria.
+func MatchesFinegrainedException(
+	polexs []*kyvernov2.PolicyException,
+	policyContext engineapi.PolicyContext,
+	policyName, ruleName string,
+	logger logr.Logger,
+) ([]kyvernov2.PolicyException, kyvernov2.ExceptionReportMode) {
```

Comment:
The logic should be implemented [here](https://github.com/kyverno/kyverno/blob/8b26a1ce076d80ec128fd7ff011acc7e6d370174/pkg/engine/utils/exceptions.go#L18).